sympy/ntheory/generate.py

```diff
@@ -41,8 +44,47 @@ class Sieve:
     array('L', [2, 3, 5, 7, 11, 13, 17, 19, 23])
     """
 
+    #
+    # Notes about thread safety.
+    #
+    # The global sieve instance is used indirectly by many high-level sympy
+    # functions and needs to work correctly under multithreading. Managing
+    # shared mutable state like this is tricky and requires careful attention
+    # to exactly when and how it is accessed.
+    #
+    # The pattern used here is known as Read, Copy, Update (RCU) and allows
+    # fast read access without the use of locks for synchronisation.
```

Comment:
I wrote a PR to document this better in the free-threading guide: https://github.com/Quansight-Labs/free-threaded-compatibility/pull/302. You can maybe link there once that is merged?

Reply:
Yes, that looks good.